### PR TITLE
Script for Zeta+ACA test

### PR DESCRIPTION
**Pull request by PikaPikaW**
Script for Zeta+ACA test

### CODE REVIEW

**Comment by er1cthe0ne** [comment]
@PikaPikaW great start towards #163 , it answered some of the highlevel technical questions for the setup. Adding @zzxgzgz @liangbin-pub to review.

**Comment by zzxgzgz** [comment]
Great job.

FYI, you can refer to [this page](https://github.com/futurewei-cloud/zeta/blob/main/docs/design/zeta_system_design.md#73-external-apis) for what data to POST to Zeta.

**Comment by PikaPikaW** [comment]
hello,everyone.Now I want to add a function in gtest that parses the JSON file from the controller, but I'm not sure which CPP file I should add it in, in aca_test_ovs_util.cpp? or something else; And should I define a parser function with TEST or a normal function, or should I define a TEST function which call the parser function, because when RUN_ALL_TESTS() runs, I think it will start executing all the functions with TEST, but I don't know in what order?

**Comment by er1cthe0ne** [comment]
> hello,everyone.Now I want to add a function in gtest that parses the JSON file from the controller, but I'm not sure which CPP file I should add it in, in aca_test_ovs_util.cpp? or something else; And should I define a parser function with TEST or a normal function, or should I define a TEST function which call the parser function, because when RUN_ALL_TESTS() runs, I think it will start executing all the functions with TEST, but I don't know in what order?

Hi, you will define all the zeta test under /test/gtest/aca_test_zeta_programming.cpp, you can do the following:

add new utility function: void aca_test_zeta_setup - parse JSON file from controller, setup the goal state and push to aca (Aca_Comm_Manager::get_instance().update_goal_state) - see this example: aca_test_1_neighbor_CREATE_DELETE

add new test case: TEST(zeta_programming_test_cases, gateway_path) 
add new test case: TEST(zeta_programming_test_cases, direct_path)
and the above two test case will call aca_test_zeta_setup in the begining then execute the test

**Comment by PikaPikaW** [comment]
What do the fields IP and VIP in data mean when adding port?
```shell
Request: http://localhost:8080/ports
    data:
    [
      {
        "port_id": "333d4fae-7dec-11d0-a765-00a0c9342222",
        "vpc_id": "3dda2801-d675-4688-a63f-dcda8d327f50",
        "ips_port": 
        [
          {
            "ip": "10.10.0.3",
            "vip": ""
          },
          {
            "ip": "10.10.3.7",
            "vip": "10.10.3.100"
          }
        ],
        "mac_port": "cc:dd:ee:ff:11:22",
        "ip_node": "192.168.10.27",
        "mac_node": "ee:dd:ee:ff:22:11",
      },
      {
        "port_id": "99976feae-7dec-11d0-a765-00a0c9341111",
        "vpc_id": "3dda2801-d675-4688-a63f-dcda8d327f55",
        "ips_port": 
        [
          {
            "ip": "10.10.0.3",
            "vip": ""
          },
          {
            "ip": "10.10.3.7",
            "vip": ""
          }
        ],
        "mac_port": "6c:dd:ee:ff:11:32",
        "ip_node": "192.168.10.33",
        "mac_node": "ee:dd:ee:ff:33:11",
      }
    ]
```

**Comment by er1cthe0ne** [comment]
>     ]
> ```

just use this:

        "ips_port": 
        [
          {
            "ip": "10.10.0.3",
            "vip": ""
          }
        ],

**Comment by er1cthe0ne** [comment]
check out below to see if there is other json parser which is easier to use.

https://stackoverflow.com/questions/36471581/c-c-json-parser

**Comment by zzxgzgz** [comment]
I also made a fork and tried to update the json library && fix compile errors in `aca_test_zeta_programming.cpp`, you can take a look and pick it up from there:
https://github.com/zzxgzgz/alcor-control-agent/tree/integrationTest

**Comment by PikaPikaW** [comment]
Ok, thanks for your help very much, so use 'nlohmann/Json.hpp' to replace 'jsoncpp' to parse json file, commit will not fail,or i shall add nlohmann's source code in aca?

**Comment by PikaPikaW** [comment]
What do the fields gws in responsed data mean when adding vpc? need to put in configuration file?

```shell
data:
    {
        "vpc_id": "3dda2801-d675-4688-a63f-dcda8d327f50",
        "vni": "12345",
        "zgc_id": "f81d4fae-7dec-11d0-a765-00a0c91e6bf6",
        "name": "ZGC_test1",
        "gws": 
        [
          {
            "ip": "192.168.0.87",
            "mac": "37.02.ff.cc.65.87"
          },
          {
            "ip": "192.168.0.88",
            "mac": "37.02.ff.cc.65.88"
          },
          {
            "ip": "192.168.0.89",
            "mac": "37.02.ff.cc.65.89"
          }
        ],
        "port_ibo": "8300"
    }
```

**Comment by liangbin-pub** [comment]
gws stands for gateways, those are the list of ZGC gateway IP/MAC pairs to populate OVS group/bucket

**Comment by zzxgzgz** [comment]
> Ok, thanks for your help very much, so use 'nlohmann/Json.hpp' to replace 'jsoncpp' to parse json file, commit will not fail,or i shall add nlohmann's source code in aca?

The `nlohmann/Json.hpp` is already in /include/nlohmann/Json.hpp in my branch integrationTest branch. You can pull from my repo if you want to.

**Comment by er1cthe0ne** [comment]
> What do the fields gws in responsed data mean when adding vpc? need to put in configuration file?

The fields that's needed to construct goal state to ACA are below, they should be put in configuration file:

    {
        "zgc_id": "f81d4fae-7dec-11d0-a765-00a0c91e6bf6",
        "gws": 
        [
          {
            "ip": "192.168.0.87",
            "mac": "37.02.ff.cc.65.87"
          },
          {
            "ip": "192.168.0.88",
            "mac": "37.02.ff.cc.65.88"
          },
          {
            "ip": "192.168.0.89",
            "mac": "37.02.ff.cc.65.89"
          }
        ],
        "port_ibo": "8300"
    }

File: test/zeta-aca-test_controllerScript/run.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    requests.post(zgc_nodes_url[0] + "/ports", PORT_data)
+    requests.post(zgc_nodes_url[1] + "/ports", PORT_data)
+    # TODO: 分别生成CHILD和PARENT的配置文件
+
+
+def run():
+    # # Call zeta API to create ZGC,vpc etc.and generate the information ACA need, and save it in zetaToAca_data.json
+    file_path = './data/zeta_data.json'
+    zgc1_nodes_url = ["http://172.16.62.247:8080", "http://172.16.62.248:8080"]
+    talk_to_zeta(file_path, zgc1_nodes_url)
```
Comment (zzxgzgz) [inline test/zeta-aca-test_controllerScript/run.py:104]: This is actually not what you should do with the zgc_node_url(ip), like calling /zgc, /vpc in the `talk_to_zeta`, you also want to call the /nodes POST api for these zgc_node_url. 

You should add a few entries of data in your `zeta_data.json`:
- zeta_api_ip (where you will actually call the /zgc, /vpc, /nodes APIs)
- NODE_data, with the format that the /nodes POST api desires [(refer to here)](https://github.com/futurewei-cloud/zeta/blob/main/docs/design/zeta_system_design.md#7331-add-zgc-node)

Then, in the `talk_to_zeta`, after creating the calling /zgc and /vpc, call /nodes POST to add the zgc1_nodes_url, can you refer to [here](https://github.com/futurewei-cloud/zeta/blob/main/deploy/install/deploy_wellcheck.sh#L102) for how to make the call.

**Comment by PikaPikaW** [comment]
hello,i am a little busy recently, very sorry. i have carefully read the changed code. So, when talking to zeta,the order is creating a ZGC firstly, getting a zgc_id returned, then creating the ZGC Nodes information in this zgc_id, then creating VPC, finally adding port information, is that right?
I have some questions about several IP entries in the Zeta data:
- When creating ZGC, ip_start, IP_end refers to the IP address area that will be assigned to ZGC Node in ZGC ? I installed Zeta on machine 247 and 248 here, so the address area should be written as 172..16.62.247-248. Is that right ?
- When creating ZGC Node, is IP_contorl ZGC Node's ip address? I only have two machines install Zeta, so I should create two nodes ?
- When creating VPC, the gws returned is the ZGC Gateway's information (Is ZGC Node and ZGC Gateway a concept?).Are these IP addresses corresponding to the IP address area given when creating ZGC? and this information should be placed in the configuration file ?

**Comment by liangbin-pub** [comment]
Hi Wanchen,
1.no, the ip range given to ZGC is for ZGC gateway entrances on tenant network(vxlan), not the one for management (such as you ssh into it), they are of different network
2. No, if you have two machine to run Zeta, you will have one for Zeta management (serve the RESTful APIs), and the other as ZGC "node", which serve as gateway for tenant network traffic
3. yes, the gws list is same as the range given at ZGC create, with additional info with MAC

**Comment by zzxgzgz** [comment]
Hi @PikaPikaW,

I made some changes to the aca test code, and made some minor modifications to the pseudo controller. Please take a look and consider pulling my changes. 

Thanks

**Comment by PikaPikaW** [comment]
Okay, I think i get the idea, so:
- There will be one node in each ZGC for communication with management (restful apis), and other nodes as the gateway
- ZGC Node that communicates with Management will have multiple IPs, IP that communicates with Manangement (IP used when ssh into it), and IP that communicates with ACA (IP in Tenant Work)

If I understand correctly, then:
- The node in ZGC that is responsible for communicating with Management gets the information that needs to be added, how other nodes get information, whether they communicate with this node or also with Management (in cases script not used)?
- Ip_control is the IP that ZGC Node communicates with Management?

**Comment by PikaPikaW** [comment]
@er1cthe0ne, okey

**Comment by PikaPikaW** [comment]
okay, @zzxgzgz

**Comment by PikaPikaW** [comment]
- i think the IP_node in the responed port information should be the ip of the vm host. Why is this ip different from the Ip information in ACA_nodes

- In addition, in your experimental environment, how many Zeta gateway nodes and ACA compute nodes respectively?

**Comment by zzxgzgz** [comment]
> * i think the IP_node in the responed port information should be the ip of the vm host. Why is this ip different from the Ip information in ACA_nodes
> * In addition, in your experimental environment, how many Zeta gateway nodes and ACA compute nodes respectively?

@PikaPikaW 
- Yes and no. The ip_node should be one of the IP of the interface you use for tenant network of that host. Also, are you guys using VMs as ACA_nodes? We're using physical hosts for ACA nodes, and those nodes don't run any VMs, so I'm not sure if I'm understanding the `the ip of the vm host` correctly. 
- In our experimental environment, we have one physical node to deploy zeta and the pseudo controller, one physical node to run 6 VMs as zeta gateway nodes, one physical node to be the master of the zeta gateway cluster, and two physical nodes as ACA compute nodes.

image

**Comment by PikaPikaW** [comment]
Well, thanks for your explanation very much.so pseudo controller communicates with Zeta Control plane, adding information(step 1-4).Then 
- does ZGC Nodes need to save the added information? If need, should the master of the Zeta Gateway cluster send the information, or pseudo Contoller (in addition, the master of the Zeta Gateway cluster and ZGC nodes communicate in the ZGC_network?)
- does ZGC nodes need to install Zeta?
- Does the pseudo Controller's physical machine also need to install Zeta (I don't know what Zeta repo for Zeta Deployment means)?
- so,in node's information, ip_control is the ip of ZGC Node in tenant_network, which also corresponds to the ip area in zgc's information. However, in the figure, it seems that the ip of each ZGC Node's ens4 is not in the ip area (ip_start :192.168.20.100; Ip_end: 192.168.20.115)
- In our experimental environment, the current environment is the same as the previous installation, as shown in the figure below, which also uses two physical machines as ACA Nodes and does not run any vm.ZGC also uses two physical machines with Zeta installed and does not use any vm. If one is the master of the Zeta Gateway cluster, then there is only one ZGC Node
image

**Comment by PikaPikaW** [comment]
hello,when i am running the script,met some problem:
- After zeta is installed, there will be a default zgc information, I want to delete this zgc, but using API can not delete,i think before delete zgc,we should clean its node information,However, there is no id information of node when using get zgc.
- When adding vpc, the message sent does not include zgc_id, so which zgc ip area does the returned gws correspond to?

because I cannot delete the default ZGC, so after adding VPC, the returned gws corresponds to the IP area of the zgc that existed by default at the beginning(not the new zgc i add) I don't know how to solve this problem ?

**Comment by liangbin-pub** [comment]
can you use the node delete API first to delete nodes then delete ZGC? They have dependencies

**Comment by zzxgzgz** [comment]
@PikaPikaW I just read your latest commit, and I'd like to suggest you not to put your username/password into the json file, as now you exposed their login info and it is unsafe. Please find a way to fix this.

**Comment by PikaPikaW** [comment]
Hello, may I ask, what dependencies should I delete before I delete port. Because I want to delete vpc, I think delete vpc information needs to delete the port related to it firstly, and the port deletion always fails, prompt 500. In the information returned by GET ports, what do eps and host_id stand for? I understand from the api source code that these two fields also have GET operation, but there is no DELETE and POST operation.

**Comment by zzxgzgz** [comment]
> Hello, may I ask, what dependencies should I delete before I delete port. Because I want to delete vpc, I think delete vpc information needs to delete the port related to it firstly, and the port deletion always fails, prompt 500. In the information returned by GET ports, what do eps and host_id stand for? I understand from the api source code that these two fields also have GET operation, but there is no DELETE and POST operation.

Can you try to check the logs for the zeta-manager pods?

You can do `kubectl get pods` to find all pods in your zgc cluster, then `kubectl logs zeta_manager_pod_name` to check the logs.

**Comment by PikaPikaW** [comment]
@liangbin-pub,hello,the zgc control node has been deployed,the machine's port is  8221

**Comment by liangbin-pub** [comment]
@PikaPikaW Thanks, I checked zeta repo is good, thanks

**Comment by PikaPikaW** [comment]
Hello, I would like to confirm that now tower1 221 is used as Zeta Control Plane Server and an ACA Node. So tower2 222 is used as another ACA node or used to deploy ZGC nodes? and the other two servers 172.16.50.7 and 172.16.50.8 how to allocate?

**Comment by liangbin-pub** [comment]
No, tower 1 is used for:
ACA parent, Pseudo controller and Zeta deployer (which deploys zeta to remote servers)

Tower 2 is used for: ACA child

Server 1 is used for ZGC nodes (will deploy 6 VMs there) as ZGC nodes (172.16.150.190-172.16.150.195 on tenant network and 172.16.50.190-172.16.50.195 on management network)

Server2 is used for Zeta control plane (K8S cluster) which offers Zeta REST API @ http://172.16.50.8/

Hope this is clear

**Comment by liangbin-pub** [comment]
Update:
I lost connectivity to server 2, so I'm using server1 (50.7) to deploy ZGC nodes (VMs), it's done, total 6 VMs. 
To re-deploy (normally not needed), login and in ~/kvm, run "sudo ./kvm_deploy.sh 6", these will delete old VMs and re-deply.

Once Qianyu brings back server2 (fresh install ok), Zeta control plane services (providing REST API http://172.16.50.8/)can be deployed on it following following step:
1. login tower1 (172.16.50.221)
2. cd ~/zeta
3. ./deploy/full_deploy.sh -d lab

Then follow ACA instructions to deploy ACA on both tower1 (parent) and tower2 (child)
 
When using Pseudo controller, make sure change following parameters:
- ip_start and ip_end in ZGC create API call, give it 172.16.150.80-172.16.150.95   (16 unused IPs in 172.16.150.0/24)
- when adding AGC nodes (6 times for the 6 ZGC node VMs), ip_control is 172.16.50.190 for node1, 172.16.50.191 for node2, etc, up to node 6; id_control and pwd_control van put any string, not used; int_tenant put "ens4", inf_zgc put "ens3", you need to get MAC address of each ZGC node VM's ens3, ens4 interface and fill in as mac_zgc and mac_tenant.

**Comment by er1cthe0ne** [comment]
per 12-21-2020 open source community discussion, closing this older PR. Please rebase with the latest code and create new PR for the next code change. https://github.com/futurewei-cloud/alcor-control-agent/blob/master/src/README.md

**Pull request closed without merging**